File: cpp/cpp_100-999/636.cc
// C++
// leetcode 636
// https://leetcode.cn/problems/exclusive-time-of-functions/

class Solution {
public:
    vector<int> exclusiveTime(int n, vector<string>& logs) {
        vector<int> res(n, 0);
        vector<int> cpu;
        int cur = -1;

        for (auto &s : logs) {
            // 解析日志 s
            int nn = s.size(), p = 0;
            while(s[p] != ':')
                p++;
            int idx = stoi(s.substr(0, p++));
            char c = s[p];
            while(s[p] != ':')
                p++;
            int val = stoi(s.substr(++p));

            // 入
            if (c == 's') {
                // 原本非空
                if (!cpu.empty()) {
                    res[cpu.back()] += val - cur;
                }
                cpu.push_back(idx);
            } else {
            // 出
                val++;
                res[cpu.back()] += val - cur;
                cpu.pop_back();
            }
            cur = val;
        }

        return res;
    }
};

File: Go/Go_100-999/636.go
// Go
// leetcode 636
// https://leetcode.cn/problems/exclusive-time-of-functions/

import (
	"strconv"
)

func exclusiveTime(n int, logs []string) []int {
	res := make([]int, n)
	cur := 0
	stk := make([]int, 0, 4)

	for _, s := range logs {
		p := 0
		for s[p] != ':' {
			p++
		}
		idx, _ := strconv.Atoi(s[0:p])

		p++
		c := s[p]
		for s[p] != ':' {
			p++
		}
		val, _ := strconv.Atoi(s[p+1:])
		// idx c val

		// 入
		sz := len(stk)
		if c == 's' {
			if sz != 0 {
				res[stk[sz-1]] += val - cur
			}
			stk = append(stk, idx)
		} else {
			// 出
			val++
			res[stk[sz-1]] += val - cur
			stk = stk[0 : sz-1]
		}

		cur = val
	}

	return res
}